Add CLOCK based approximate LRU

JLockerman opened the pull request:
We have a number of places in the connector where we want to store a
cache of elements. Our expectation is that the set of live elements
will be time-correlated, and therefore, when we want to evict elements
from the cache the least-recently-used element is the one least likely
to be used again in the future. There for this commit adds an LRU
implementation suitable for use in such caches. Traditional LRUs suffer
from large amounts of contention in multi-threaded workloads, so we use
a CLOCK-based approximate LRU; more information can be found in the
README.

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+	used uint32
+
+	// pad Elements out to be cache aligned
+	_padding [24]byte
```
`_padding` should be named `_`

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+	}
+}
+
+func (self *Cache) tryEvict() (insertPtr *Element, evicted bool) {
```
can't we rewrite cache.evict() as the following and avoid the try. The following should guarantee we always return
```
for {
      if self.next >= len(self.storage) {
		self.next = 0
      }
      elem := &self.storage[self.next]
      if elem.used != 0 {
         elem.used = 0
      }
      else {
	       delete(self.elements, elem.key)
               return elem
      }
      self.next += 1
}
```

iand replied:
Think this would need to increment self.next before returning elem

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+	return elem.Value, true
+}
+
+func (self *Cache) Unmark(key string) bool {
```
Wouldn't a Remove be more appropriate? What's the use-case for just an Unmark where you wouldn't want a remove?

JLockerman replied:
1. Testing 😋 
2. incremental unmarking: one of the big downsides to this kind of cache is the ~linear time it takes to evict. One thing that may help with that is incrementally unmarking elements during normal operation so that when it comes time to evict there's less to do. `Remove` is much to aggressive an operation for that.

cevian commented on pkg/clockcache/cache_test.go:
```diff
@@ -0,0 +1,173 @@
+package clockcache
```
This would be a great use case for some testing benchmarking functions too

JLockerman replied:
hmm, I have some benchmarks I copied of `bigcache` that I've been using, but I don't love them. It's a bit difficult to get benchmarks that are indicative of real-world contention, which is what should really matter for performance. I'll see what I can do...

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+
+// CLOCK based approximate LRU storing mappings from strings to
+// int64s designed for concurrent usage.
+// It is sharded into 127 blocks, each of which can be
```
comment needs updating; no longer sharded.

iand commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+// block. Eviction locks blocks one at a time looking for a value
+// that's valid to evict/
+type Cache struct {
+	lock sync.RWMutex
```
This lock protects access to `elements`+`storage` and `next`. I suggest having a separate Mutex for `storage`+`next`. This allows cache reads to take a read lock without having to wait for a pending eviction from a concurrent insert.

RWMutexes have a pathological case where every call to Lock causes all RLocks to be queued until Unlock is called.

JLockerman replied:
@iand you mean something like
```
1. lock storage+next, search for an element to be evicted
2. lock elements, evict old element and re-insert a new one
```
?
I'll have to think about that a bit more...

JLockerman replied:
one issue is that it can cause the evictor to starve if the other threads are hitting the cache fast enough...

iand replied:
Yes, that occured to me while I was out for my walk. If the cache is hot such that every key is being used between writes then it's probably too small and inserting will just cause unnecessary churn.  However, is there a requirement for insert to always succeed? It could attempt one run through the eviction loop and return if it can't insert. On a hot cache that's not much different from inserting and being immediately evicted.

JLockerman replied:
That's a good point, and in the current implementation `InsertBatch` can actually "succeed" by evicting its own elements, which is morally the same thing.

JLockerman replied:
@iand @cevian Ok, let;s go through this in more depth.
The evict we use in `Insert` will go through `storage` until we either find an element to evict or we reach the element we started at without having done so. In the second case we’ll give up, as all the elements are too useful (maybe we’ll add a new heuristic later, really this probably means the cache is too small). I’m going to assume it’s ok to be a little sloppy with `used`, and that a sequence like
```
1. evictor sees Element e is unused, decides to evict it
2. getter marks e as used
3. evictor grabs lock on elements and evicts e
```
is ok: it’s morally equivalent to swapping the order of steps `1` and `2`.

The changed functions are
```go
func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
	elem, present := self.elements[key]
	if present {
		return elem.key, elem.Value, false
	}

	var insertLocation *Element
	if len(self.storage) >= cap(self.storage) {
		insertLocation = self.evict()
		if insertLocation == nil {
			return nil, nil, false
		}
		self.elementsLock.Lock()
		defer self.elementsLock.Unlock()
		key := insertLocation.key
		delete(self.elements, key)
		*insertLocation = Element{key: key, Value: value}
	} else {
		self.storage = append(self.storage, Element{key: key, Value: value})
		insertLocation = &self.storage[len(self.storage)-1]
	}

	self.elements[key] = insertLocation
	return key, value, true
}

func (self *Cache) evict() (insertPtr *Element) {
	for {
		if self.next >= len(self.storage) {
			self.next = 0
		}

		insertPtr = self.tryEvictNext()
		if insertPtr != nil {
			return
		}
		if self.next == startLock {
			break
		}
	}
	// we've gone through all the elements, and they're all used
	// we'll try going through them once more, and if we cannot find
	// an unused element, the cache is to small, and we'll give up
	for {
		if self.next >= len(self.storage) {
			self.next = 0
		}

		insertPtr = self.tryEvictNext()
		if insertPtr != nil {
			return
		}
		if self.next == startLock {
			break
		}
	}
	return
}

func (self *Cache) tryEvictNext() (insertPtr *Element) {
	elem := &self.storage[self.next]
	used := atomic.CompareAndSwapUint32(&elem.used, 1, 0)
	if !used {
		insertPtr = elem
	}
	self.next += 1
	return
}

```

JLockerman replied:
so:

1. The long-duration lock only excludes other evictions, which is really nice; the only place we lock out `Get`s is for a `delete;insert` into the `map`
2.Correctness looks pretty trivial: the only state touched by the evictor is `next`, which is guarded by the lock, and `used` which is only accessed via atomics.
3. We can _almost_ forgo locking during the eviction-search entirely, if we use atomic ops to update `next` there're no race conditions, though there is a little weirdness when to evictors decide to evict the same element if one loops the other. I'm still tempted tho, I think that case should be rare enough, and the extra performance would be worth it. (this can be mitigated to a large degree by css-ing `2` into `used` when deciding to evict, it's not perfect because the `Get` update of `used` is non-atomic, but I'm not sure we care)
3. CPU-cache behavior is roughly the same: everyone has an oblige cache-miss to grab the lock, the evictor has an obligate miss per-element searched through, and however much it does when manipulating the map; the only addition is an extra lock-acquisition

JLockerman replied:
@iand nice find! if you and @cevian can confirm that the above code looks right (modulo the lock changes in `Insert[Batch]` which I elided) I'll update the PR to match

JLockerman replied:
writing this down so I don't forget: non-locking eviction has an actual data-race where the
```go
*insertLocation = Element{key: key, Value: value}
```
races against 
```go
used := atomic.CompareAndSwapUint32(&elem.used, 1, 0)
```
I think this is "benign" in that both are writing the same value, and I don't know of an architecture that would mess it up, but we still shouldn't do it.

cevian replied:
A few questions:
1) what lock protects next? I don't see it in the code above.
2) why is the Get update to Used non-atomic?

cevian replied:
Also what are the locks taken by Get?

JLockerman replied:
@cevian 
> what lock protects next? I don't see it in the code above.
Yeah, that's the one I elided 😊 . There's a mutex that protects next, and is acquired [here](https://github.com/timescale/timescale-prometheus/pull/147/files#diff-666b355b2c2f901c37b3cccc26156de4R52-R53l) and [here](https://github.com/timescale/timescale-prometheus/pull/147/files#diff-666b355b2c2f901c37b3cccc26156de4R67-R68) instead of the current lock.

> why is the Get update to Used non-atomic?
we don't have enough words. It's non-atomic because it uses
```go
if atomic.LoadUint32(&elem.used) == 0 {
    atomic.StoreUint32(&elem.used, 0)
}
```
instead of 
```go
atomic.CompareAndSwap(...)
```
for reasons described [here](https://github.com/timescale/timescale-prometheus/pull/147/files#r453666623), and that means the `insert` update to `used` can occur between the `if` and the `Store`. (It still is atomic in the sense that it uses atomic operations to shared memory, and does not have data-races)

JLockerman replied:
> Also what are the locks taken by Get?

@cevian The exact same locks that are taken now; the `Get` code should not need to change.

iand replied:
This looks better to me

iand commented on pkg/clockcache/Readme.md:
```diff
@@ -0,0 +1,72 @@
+
+## Other Optimizations ##
+
+The cache is built off a sharded hashmap; values are first hashed to see which
```
Needs removing now

iand commented on pkg/clockcache/Readme.md:
```diff
@@ -0,0 +1,72 @@
+and the write side for cache inserts and evictions. Eviction locks a single
+shard at a time, going in round-robin fashion until it finds an element to evict.
+
+Each shard consists of a hashmap and linkedlist. The map is used to perform key
```
It's not a linked list, just a list.

JLockerman commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+	if len(keys) != len(values) {
+		panic(fmt.Sprintf("keys and values are not the same len. %d keys, %d values", len(keys), len(values)))
+	}
+	values = values[:len(keys)]
```
it's mostly a lint; if you forget a key or a slot for a value it's really annoying to debug, and it's not difficult to reslice before calling `InsertBatch` when you want to store into a longer sice

JLockerman commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+
+// tries to get a batch of keys and store the corresponding values is valuesOut
+// returns the number of keys that were actually found.
+// NOTE: this function does _not_ preserve the order of keys; the first numFound
```
The real reason is because the API was designed around non-pointer keys. However, the current version has the advantage of constant-time discovery of hits/misses. In our labels cache, the expected use looks something like:
```go
numFound := labelsCache.GetValue(labelIds, labels)
if numFound < len(labelIds) {
    missed := fetchLabels(labelIds[numFound:])
    copy(labels[numFound:], missed)
}
sort.Slice(labels, ...)
return labels
```
which I think lends itself better to the current API

JLockerman commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+		return 0, false
+	}
+
+	if atomic.LoadUint32(&elem.used) == 0 {
```
unless we do the changes Ian proposes around multiple locks, it shouldn't: the lock will exclude the evictor, and all readers are going to write the same value at this point, so the store is idempotent. An atomic store is potentially a decent bit faster than a `cmpxchg`, so we want to use the simplest instruction we can

JLockerman replied:
I assumed you meant the next line down, but I realize now you may mean the while
```go
if atomicLoad(...)  {
    atomicStore(...)
}
```
The current way actually performs a bit better: this is a read-mostly operation; the majority of access to the cache should be `Get`s, which means most of the time when we reach this code `used` will already be `1`, and we'll never hit the store. If we perform a `CompareAndSwap` we'll store a value every single time we perform a `Get`. Performing a read to a shared CPU cache-line is basically free (if it's already in the cache), while performing a store require gaining exclusive access to the line, both increasing the latency for the store-er, and forcing a cache-miss on all other cores.

JLockerman replied:
I'l add a comment summarizing this.

JLockerman commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+		return false
+	}
+
+	if atomic.LoadUint32(&elem.used) != 0 {
```
if you're talking about the entire `if` then see https://github.com/timescale/timescale-prometheus/pull/147/files#r453666623.
If you're worried about races, then there are two cases:

1. A `Get` sets used to `1` after we perform the `Load`. This doesn't matter, since the value remains the same
2. Another `Unmark` sets the value to `0` after the `Load`. Both threads write the same value, so it's idempotent.

Notably the evictor cannot be running because of the lock.

JLockerman commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,201 @@
+
+type Element struct {
+	// The value stored with this element.
+	key   interface{}
```
Yeah, I don't love this. The problem we have is that we want to use both `int64`s and `string`s as keys, and I'm not sure @cevian is willing for us to `generate` a unique cache type for each. I considered storing redundant key and fields for different types, but since we'd need to `interface{}` the keys anyway to store in the `map`, I don't think it's worthwhile.
(`interface{}`ing will _also_ increase the cache miss rate normally, because of the pointer-chasing you mentioned, the padding below is to prevent false-sharing and bouncing elements around cores because of new adjacent elements. I'm not sure it's worthwhile in this case, because we should be read-mostly anyway, and I have concerns about the `map`, but ehh, unless we have more stuff to store here, it doesn't really matter)

iand commented:
A stylistic nit: conventionally the receiver in Go is a single character, e.g. `c` for cache. `self` stands out as odd.

JLockerman commented:
I pushed a commit with @iand's edits to the locking.

JLockerman commented:
I pushed a new version of the eviction code that eliminates bounds checking.
@iand I pushed a version of your benchmarks. I ended up changing them slightly to ensure that evictions happened.

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,289 @@
+	"sync/atomic"
+)
+
+// CLOCK based approximate LRU storing mappings from strings to
```
the mapping is between interfaces right not strings and int64s

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,289 @@
+	return len(keys)
+}
+
+func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
```
```suggestion
func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, in_cache bool) {
```

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,289 @@
+	return len(self.storage)
+}
+
+func (self *Cache) DebugString() string {
```
we should either delete this or make debugString

cevian commented on pkg/clockcache/cache.go:
```diff
@@ -0,0 +1,289 @@
+
+	newStorage := make([]element, 0, newMax)
+
+	// can we use copy here despite the data race
```
```suggestion
	// cannot use copy here because of the data race on element.used
```

JLockerman commented:
benchmark results on a standard droplet
```
goos: linux
goarch: amd64
pkg: github.com/timescale/timescale-prometheus/pkg/clockcache
BenchmarkIntCache-8              	 4132209	       314 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringCache-8           	 3913479	       341 ns/op	       0 B/op	       0 allocs/op
BenchmarkEviction/500-8          	 1377842	       862 ns/op	       0 B/op	       0 allocs/op
BenchmarkEviction/5000-8         	  165669	      7369 ns/op	       0 B/op	       0 allocs/op
BenchmarkEviction/50000-8        	  191414	     70761 ns/op	       0 B/op	       0 allocs/op
BenchmarkMembership/500-8        	20423635	        58.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkMembership/5000-8       	17888761	        65.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkMembership/50000-8      	12135523	        96.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkMembership/500000-8     	 6010875	       198 ns/op	       0 B/op	       0 allocs/op
BenchmarkNotFound/500-8          	24013395	        51.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkNotFound/5000-8         	20420546	        58.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkNotFound/50000-8        	17699750	        67.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkNotFound/500000-8       	13643972	        83.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkInsertUnderCapacity/500-8         	 7455434	       162 ns/o   7455434 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertUnderCapacity/5000-8        	 7307496	       164 ns/o   7307496 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertUnderCapacity/50000-8       	 6761834	       179 ns/o   6761834 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertUnderCapacity/500000-8      	 4576710	       233 ns/o   4576710 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertOverCapacity/500-8          	 5698213	       211 ns/o   5698213 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertOverCapacity/5000-8         	 6743119	       177 ns/o   6743119 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertOverCapacity/50000-8        	 7207106	       167 ns/o   7207106 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertOverCapacity/500000-8       	 6356412	       175 ns/o   6356412 inserts	         1.00 inserts/ops	       0 B/op	       0 allocs/op
BenchmarkInsertConcurrent/500-8            	 4390713	       280 ns/o       0 B/op	       0 allocs/op
BenchmarkInsertConcurrent/5000-8           	 3873616	       296 ns/o       0 B/op	       0 allocs/op
BenchmarkInsertConcurrent/50000-8          	 4385242	       317 ns/o       0 B/op	       0 allocs/op
BenchmarkInsertConcurrent/500000-8         	 3538814	       373 ns/o       0 B/op	       0 allocs/op
```

iand commented:
The insert counts on the benchmark look incorrect. 1 per op means every insert was new or evicted an existing entry. 

For comparison, this is what I saw when I originally wrote them in my branch:

```
BenchmarkInsertUnderCapacity/500-8         	12273985	        98.5 ns/op	         0.000030 inserts/op
BenchmarkInsertUnderCapacity/5000-8        	12334939	        97.7 ns/op	         0.000229 inserts/op
BenchmarkInsertUnderCapacity/50000-8       	11569123	       101 ns/op	         0.00194 inserts/op
BenchmarkInsertUnderCapacity/500000-8      	 8644075	       131 ns/op	         0.0201 inserts/op
BenchmarkInsertOverCapacity/500-8          	 9168835	       131 ns/op	         0.783 inserts/op
BenchmarkInsertOverCapacity/5000-8         	 9761977	       128 ns/op	         0.596 inserts/op
BenchmarkInsertOverCapacity/50000-8        	 9785414	       123 ns/op	         0.458 inserts/op
BenchmarkInsertOverCapacity/500000-8       	10211001	       135 ns/op	         0.350 inserts/op
```

The UnderCapacity benchmark shows that repeated iterations of the benchmark are almost always hitting a pre-existing key, which is what you expect if the cache is large enough to cover the test data.

The OverCapacity benchmark shows lots of evictions since the cache is only 25% of the size of the test data set.

JLockerman commented:
Ah whoops. I changed the meaning of the `bool` returned by `Insert` a never updated the benchmarks 😅 
(for those of you coming in from the future, at one point `Insert` returned `true` if the call to the function inserted the value into the cache and `false` if the value was already there and we're returning the old value. At the last minute I changed it to return `true` if the value is now in the cache and `false` if the `Insert` starved. My logic is that concurrent `Insert`s should be trying to insert the same value so it's not very useful to know which version inserted, you just always use the one returned, on the other hand, it can be very useful to know if the cache is overloaded)
